### PR TITLE
Make email name quoting more complete

### PR DESCRIPTION
**Pull request by dscho**
This PR tries to address the shortcomings pointed out in https://github.com/gitgitgadget/gitgitgadget/pull/98#issuecomment-506617659.

### CODE REVIEW

File: lib/patch-series.ts [hunk 1]
```diff
@@ -328,23 +328,19 @@ export class PatchSeries {
 
     protected static encodeSender(sender: string): string {
         const encoded = encodeWords(sender);
-        const dot = encoded.indexOf(".");
-        if (dot < 0) {
+
+        /* Don't quote if already quoted */
+        if (encoded.startsWith("\"") && encoded.indexOf("\" <") > 0) {
```
Comment (sunshineco) [inline lib/patch-series.ts]: I would have expected this to match whitespace loosely (just like the check later in the patch for a _special_ character in the name):

```
    ... && encoded.match(/"\s*</) {
```

File: lib/patch-series.ts [hunk 2]
```diff
@@ -328,23 +328,19 @@ export class PatchSeries {
...
-        if (endOfName < 0) {
-            endOfName = encoded.length;
-        } else {
-            while (endOfName > 0 && encoded[endOfName - 1] === " ") {
-                endOfName--;
-            }
-        }
-        if (dot >= endOfName) {
+        const match =
+            encoded.match(/^([^<]*[()<>[\]:;@\\,."][^<]*?)(\s*)(<.*)/);
```
Comment (sunshineco) [inline lib/patch-series.ts]: This is a bit difficult to reason about but I _think_ it's correct. To make it more obvious that this works as intended, perhaps in addition to the test @derrickstolee suggested, test cases could be added which verify that these do not get quoted:

```
name <email>
my name <email>
excited! <email> 
```

and these get quoted:

```
mr. name <email>
foo [bar] name <email>
bob@obo <email>
blug <nobody <email>
```

(I'm not sure the last one has any value; it almost certainly isn't even legal so perhaps it doesn't deserve to be tested.)

File: lib/patch-series.ts [hunk 3]
```diff
@@ -328,23 +328,19 @@ export class PatchSeries {
...
+        if (!match) {
             return encoded;
         }
-        return `"${encoded.substr(0, endOfName)}"${encoded.substr(endOfName)}`;
+
+        return `"${match[1].replace(/"/g, "\\\"")}"${match[2]}${match[3]}`;
```
Comment (sunshineco) [inline lib/patch-series.ts]: Escaping embedded double quotes make sense. Do we also need to escape backslashes?

File: lib/patch-series.ts (hunk 1 above)
Comment (dscho) [reply lib/patch-series.ts]: Makes sense! Thanks.

File: lib/patch-series.ts (hunk 2 above)
Comment (dscho) [reply lib/patch-series.ts]: I added all of them, and also one containing a backslash.

Comment (dscho) [reply lib/patch-series.ts]: I added all of them, and also one containing a backslash.

Comment (dscho) [reply lib/patch-series.ts]: I added all of them, and also one containing a backslash.

File: lib/patch-series.ts (hunk 3 above)
Comment (dscho) [reply lib/patch-series.ts]: Yes, I think so. I added it. And a test.

**Comment by dscho** [comment]
> Do you have a test to check the "Don't quote if already quoted" logic?

Now I do ;-)

**Comment by dscho** [comment]
> Do you have a test to check the "Don't quote if already quoted" logic?

Now I do ;-)

**Pull request merged**